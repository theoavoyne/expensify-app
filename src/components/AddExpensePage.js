import React from 'react';

const AddExpensePage = () => (
  <div>
    This is from AddExpensePage
  </div>
);

export default AddExpensePage;
