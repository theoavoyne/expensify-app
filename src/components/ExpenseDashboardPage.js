import React from 'react';

const ExpenseDashboardPage = () => (
  <div>
    This is from ExpenseDashboardPage
  </div>
);

export default ExpenseDashboardPage;
