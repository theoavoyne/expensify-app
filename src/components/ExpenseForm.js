import React from 'react';
import DatePicker from 'react-datepicker';
import moment from 'moment';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      note: props.expense ? props.expense.note : '',
      error: ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt: moment(createdAt) }));
    }
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }

  onFormSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  }

  render() {
    return (
      <form
        action=""
        onSubmit={this.onFormSubmit}
        className="form"
      >
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="Description"
          value={this.state.description}
          onChange={this.onDescriptionChange}
          className="text-input"
        />
        <input
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
          className="text-input"
        />
        <DatePicker
          selected={this.state.createdAt.toDate()}
          onChange={this.onDateChange}
          className="datepicker"
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
          className="textarea"
        />
        <div>
          <button type="submit" className="button">Save Expense</button>
        </div>
      </form>
    );
  }
}
