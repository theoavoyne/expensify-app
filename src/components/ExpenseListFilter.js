import React from 'react';
import { connect } from 'react-redux';
import DatePicker from 'react-datepicker';
import moment from 'moment';
import {
  sortByAmount,
  sortByDate,
  setTextFilter,
  setStartDate,
  setEndDate
} from '../actions/filters';

class ExpenseListFilter extends React.Component {
  onStartDateChange = (date) => {
    if (!date) {
      this.props.dispatch(setStartDate(null));
    } else if (date <= this.props.filters.endDate || !this.props.filters.endDate) {
      const dateMoment = moment(date);
      this.props.dispatch(setStartDate(dateMoment));
    }
  }

  onEndDateChange = (date) => {
    if (!date) {
      this.props.dispatch(setEndDate(null));
    } else if (date >= this.props.filters.startDate || !this.props.filters.startDate) {
      const dateMoment = moment(date);
      this.props.dispatch(setEndDate(dateMoment));
    }
  }

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={(e) => this.props.dispatch(setTextFilter(e.target.value))}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={(e) => {
            const { value } = e.target;
            if (value === 'date') {
              this.props.dispatch(sortByDate());
            } else if (value === 'amount') {
              this.props.dispatch(sortByAmount());
            }
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DatePicker
          selected={this.props.filters.startDate ? this.props.filters.startDate.toDate() : null}
          onChange={this.onStartDateChange}
          maxDate={this.props.filters.endDate ? this.props.filters.endDate.toDate() : null}
          isClearable
        />
        <DatePicker
          selected={this.props.filters.endDate ? this.props.filters.endDate.toDate() : null}
          onChange={this.onEndDateChange}
          minDate={this.props.filters.startDate ? this.props.filters.startDate.toDate() : null}
          isClearable
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters
});

export default connect(mapStateToProps)(ExpenseListFilter);
