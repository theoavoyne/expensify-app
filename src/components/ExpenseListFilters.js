import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import DatePicker from 'react-datepicker';
import moment from 'moment';
import {
  sortByAmount,
  sortByDate,
  setTextFilter,
  setStartDate,
  setEndDate
} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  onStartDateChange = (date) => {
    if (!date) {
      this.props.setStartDate(null);
    } else if (date <= this.props.filters.endDate || !this.props.filters.endDate) {
      const dateMoment = moment(date);
      this.props.setStartDate(dateMoment);
    }
  }

  onEndDateChange = (date) => {
    if (!date) {
      this.props.setEndDate(null);
    } else if (date >= this.props.filters.startDate || !this.props.filters.startDate) {
      const dateMoment = moment(date);
      this.props.setEndDate(dateMoment);
    }
  }

  onTextChange = (e) => (
    this.props.setTextFilter(e.target.value)
  );

  onSortChange = (e) => {
    const { value } = e.target;
    if (value === 'date') {
      this.props.sortByDate();
    } else if (value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              value={this.props.filters.text}
              onChange={this.onTextChange}
              placeholder="Search expenses"
            />
          </div>
          <div className="input-group__item">
            <select
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
              className="select"
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DatePicker
              selected={this.props.filters.startDate ? this.props.filters.startDate.toDate() : null}
              onChange={this.onStartDateChange}
              maxDate={this.props.filters.endDate ? this.props.filters.endDate.toDate() : null}
              isClearable
              className="datepicker"
            />
            <DatePicker
              selected={this.props.filters.endDate ? this.props.filters.endDate.toDate() : null}
              onChange={this.onEndDateChange}
              minDate={this.props.filters.startDate ? this.props.filters.startDate.toDate() : null}
              isClearable
              className="datepicker"
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => (
  bindActionCreators({
    setTextFilter,
    sortByDate,
    sortByAmount,
    setStartDate,
    setEndDate
  }, dispatch)
);

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
