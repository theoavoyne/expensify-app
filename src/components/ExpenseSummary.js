import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';

import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expensesTotal';

export const ExpenseSummary = ({ expensesCount, expensesTotal }) => {
  const expensePluralized = expensesCount === 1 ? 'expense' : 'expenses';
  const totalFormatted = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing&nbsp;
          <span>{expensesCount}</span>&nbsp;
          {expensePluralized} totalling&nbsp;
          <span>{totalFormatted}</span>
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = ((state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expensesCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  };
});

export default connect(mapStateToProps)(ExpenseSummary);
