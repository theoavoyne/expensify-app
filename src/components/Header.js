import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { startLogout } from '../actions/auth';

export const Header = (props) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/dashboard">
          <h1>Expensify</h1>
        </Link>
        <button
          type="submit"
          onClick={props.startLogout}
          className="button button--link"
        >
          Logout
        </button>
      </div>
    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => (
  bindActionCreators({ startLogout }, dispatch)
);

export default connect(undefined, mapDispatchToProps)(Header);
