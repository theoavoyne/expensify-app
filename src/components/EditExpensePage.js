import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onFormSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }

  onButtonClick = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onFormSubmit}
          />
          <button
            type="submit"
            onClick={this.onButtonClick}
            className="button button--secondary"
          >
            Remove Expense
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => (
  bindActionCreators({ startEditExpense, startRemoveExpense }, dispatch)
);

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
