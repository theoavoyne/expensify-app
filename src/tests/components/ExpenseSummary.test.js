import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseSummary } from '../../components/ExpenseSummary';

test('Should correctly render ExpenseSummary with one expense', () => {
  const wrapper = shallow(<ExpenseSummary expensesCount={1} expensesTotal={235} />);
  expect(wrapper).toMatchSnapshot();
});

test('Should correctly render ExpenseSummary with multiple expense', () => {
  const wrapper = shallow(<ExpenseSummary expensesCount={23} expensesTotal={56789} />);
  expect(wrapper).toMatchSnapshot();
});
