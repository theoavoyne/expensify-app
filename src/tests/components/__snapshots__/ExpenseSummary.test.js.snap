// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render ExpenseSummary with multiple expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        23
      </span>
       
      expenses
       totalling 
      <span>
        $567.89
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should correctly render ExpenseSummary with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $2.35
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
